habitat_baselines/rl/hrl/hierarchical_policy.py

```diff
@@ -0,0 +1,208 @@
+                skill_idx.item()
+            ].should_terminate(
+                batched_observations[batch_idx],
+                batched_rnn_hidden_states[batch_idx],
+                batched_prev_actions[batch_idx],
+                batched_masks[batch_idx],
+            )
+            batched_bad_should_terminate[batch_idx] = bad_should_terminate
+            self._call_high_level[batch_idx] = should_terminate
+
+        # Always call high-level if the episode is over.
+        self._call_high_level = torch.clamp(
+            self._call_high_level + (~masks).float().view(-1), 0.0, 1.0
+        )
+
+        # If any skills want to terminate invoke the high-level policy to get
+        # the next skill.
+        hl_terminate = torch.zeros(self._num_envs, device=prev_actions.device)
+        if self._call_high_level.sum() > 0:
+            (
+                new_skills,
+                new_skill_args,
+                hl_terminate,
+            ) = self._high_level_policy.get_next_skill(
+                observations,
+                rnn_hidden_states,
+                prev_actions,
+                masks,
+                self._call_high_level,
+            )
+
+            for new_skill_batch_idx in torch.nonzero(self._call_high_level):
+                skill_idx = new_skills[new_skill_batch_idx.item()]
+
+                skill = self._skills[skill_idx.item()]
+
+                (
+                    batched_rnn_hidden_states[new_skill_batch_idx],
+                    batched_prev_actions[new_skill_batch_idx],
+                ) = skill.on_enter(
+                    new_skill_args[new_skill_batch_idx],
+                    new_skill_batch_idx.item(),
+                    batched_observations[new_skill_batch_idx],
+                    batched_rnn_hidden_states[new_skill_batch_idx],
+                    batched_prev_actions[new_skill_batch_idx],
+                )
+            self._cur_skills = (
+                (1.0 - self._call_high_level) * self._cur_skills
+            ) + (self._call_high_level * new_skills)
+
+        # Compute the actions from the current skills
+        actions = torch.zeros(
+            self._num_envs, get_num_actions(self._action_space)
+        )
+        for batch_idx, skill_idx in enumerate(self._cur_skills):
+            action, batched_rnn_hidden_states[batch_idx] = self._skills[
+                skill_idx.item()
+            ].act(
+                batched_observations[batch_idx],
+                batched_rnn_hidden_states[batch_idx],
+                batched_prev_actions[batch_idx],
+                batched_masks[batch_idx],
+                batch_idx,
+            )
+            actions[batch_idx] = action
+
+        should_terminate = bad_should_terminate + hl_terminate
+        if should_terminate.sum() > 0:
+            # End the episode where requested.
+            for batch_idx in torch.nonzero(should_terminate):
+                logger.info(
```

Comment:
Is this needed?

Reply:
Yes, needed to properly call the stop action.